projects/com.oracle.truffle.llvm.parser/src/com/oracle/truffle/llvm/parser/LazyToTruffleConverterImpl.java

```diff
@@ -109,21 +117,83 @@ public RootCallTarget convert() {
                         method.getParameters().size(), symbols,
                         method, liveness, notNullable, dbgInfoHandler);
         method.accept(visitor);
+
+        LLVMControlFlowGraph cfg = new LLVMControlFlowGraph(method.getBlocks().toArray(new InstructionBlock[method.getBlocks().size()]));
+        try {
+            cfg.build();
+        } catch (Exception e) {
+            // typically do nothing and just bailout
+            // to the standard block-based dispatch without loop information
+            assert !cfg.isReducible();
+        }
+
         FrameSlot[][] nullableBeforeBlock = getNullableFrameSlots(frame, liveness.getNullableBeforeBlock(), notNullable);
         FrameSlot[][] nullableAfterBlock = getNullableFrameSlots(frame, liveness.getNullableAfterBlock(), notNullable);
         LLVMSourceLocation location = method.getLexicalScope();
 
         List<LLVMStatementNode> copyArgumentsToFrame = copyArgumentsToFrame(frame);
         LLVMStatementNode[] copyArgumentsToFrameArray = copyArgumentsToFrame.toArray(new LLVMStatementNode[copyArgumentsToFrame.size()]);
-        LLVMExpressionNode body = runtime.getNodeFactory().createFunctionBlockNode(frame.findFrameSlot(LLVMUserException.FRAME_SLOT_ID), visitor.getBlocks(), uniquesRegion.build(),
-                        nullableBeforeBlock, nullableAfterBlock, location, copyArgumentsToFrameArray);
 
+        // create function body node
+        LLVMExpressionNode body;
+        if (cfg.isReducible()) {
+            List<LLVMStatementNode> nodes = new ArrayList<>(visitor.getBlocks());
+            body = resolveLoops(nodes, cfg, frame, nullableBeforeBlock, nullableAfterBlock,
+                            location, copyArgumentsToFrameArray, uniquesRegion);
+        } else {
+            body = runtime.getNodeFactory().createFunctionBlockNode(frame.findFrameSlot(LLVMUserException.FRAME_SLOT_ID), visitor.getBlocks(), uniquesRegion.build(), nullableBeforeBlock,
+                            nullableAfterBlock, location, copyArgumentsToFrameArray, null);
+        }
         RootNode rootNode = runtime.getNodeFactory().createFunctionStartNode(runtime.getContext(), body, method.getSourceSection(), frame, method, source, location);
         method.onAfterParse();
 
         return Truffle.getRuntime().createCallTarget(rootNode);
     }
 
+    private LLVMExpressionNode resolveLoops(List<LLVMStatementNode> nodes, LLVMControlFlowGraph cfg, FrameDescriptor frame, FrameSlot[][] nullableBeforeBlock,
+                    FrameSlot[][] nullableAfterBlock, LLVMSourceLocation location, LLVMStatementNode[] copyArgumentsToFrameArray, UniquesRegion uniquesRegion) {
+
+        if (cfg.getCFGLoops().size() > 0) {
+            loopSuccessorSlot = frame.addFrameSlot("__LLVMLoopSuccessor", FrameSlotKind.Int);
+        }
+
+        LLVMStatementNode[] nodeArray = nodes.toArray(new LLVMStatementNode[0]);
```

Comment:
Use `LLVMStatementNode.NO_STATEMENTS` instead of explicitly allocating an array here.